File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/actors/PlayerActor.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.actors.ZLinkActorContext;
import systems.zlink.framework.actors.ZLinkActorJoinCompletion;
import systems.zlink.framework.actors.ZLinkActorJoinOperationId;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class PlayerActor implements ZLinkActor {
    private static final Logger logger = LoggerFactory.getLogger(PlayerActor.class);

    private final String actorId;
    private final ZLinkActorContext context;
    private String displayName;
    private String roomId = "";
    private String pendingRoomId;
    private final Set<ZLinkActorJoinOperationId> completedJoinOperations = new HashSet<>();
    private boolean destroyAfterEntrySpotJoin;
    private boolean disconnected;

    public PlayerActor(String actorId, ZLinkActorContext context) {
        this.actorId = actorId;
        this.context = context;
        this.displayName = actorId;
    }

    public String actorId() {
        return actorId;
    }

    @Override
    public ZLinkActorContext context() {
        return context;
    }

    public String displayName() {
        return displayName;
    }

    public void setDisplayName(String displayName) {
        this.displayName = displayName;
    }

    public void joinRoom(String roomId) {
        this.roomId = roomId;
    }

    public void trackDeferredJoin(String roomId) {
        if (pendingRoomId != null) {
            throw new IllegalStateException("a room join is already pending");
        }
        pendingRoomId = roomId;
    }

    @Override
    public CompletionStage<Void> onJoinCompleted(ZLinkActorJoinCompletion completion) {
        ZLinkActorJoinOperationId operationId =
                completion instanceof ZLinkActorJoinCompletion.Accepted accepted
                        ? accepted.operationId()
                        : completion instanceof ZLinkActorJoinCompletion.Rejected rejected
                                ? rejected.operationId()
                                : ((ZLinkActorJoinCompletion.Failed) completion).operationId();
        if (!completedJoinOperations.add(operationId)) {
            return CompletableFuture.completedFuture(null);
        }

        String matchedRoomId = pendingRoomId;
        pendingRoomId = null;
        if (!(completion instanceof ZLinkActorJoinCompletion.Accepted accepted)) {
            return CompletableFuture.completedFuture(null);
        }

        Messages.BingoRoomJoinRes joined = accepted.reply().decode(Messages.BingoRoomJoinRes.class);
        if (matchedRoomId == null || matchedRoomId.isBlank()) {
            matchedRoomId = joined.getState().getRoomId();
        }
        joinRoom(matchedRoomId);
        logger.info("bingo-lifecycle entry-leave actor={}", actorId);
        return CompletableFuture.completedFuture(null);
    }

    public String roomId() {
        return roomId;
    }

    public boolean destroyAfterEntrySpotJoin() {
        return destroyAfterEntrySpotJoin;
    }

    public void markForDestroyAfterRoomLeave() {
        destroyAfterEntrySpotJoin = true;
    }

    public boolean disconnected() {
        return disconnected;
    }

    public void markDisconnected() {
        disconnected = true;
    }

    // --8<-- [start:doc-bingo-bound-push]
    public CompletionStage<Void> push(Object message) {
        return context.boundSession().send(message).submit();
    }
    // --8<-- [end:doc-bingo-bound-push]
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/actors/PlayerActor.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import org.slf4j.LoggerFactory
import systems.zlink.framework.actors.ZLinkActor
import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.actors.ZLinkActorJoinCompletion
import systems.zlink.framework.actors.ZLinkActorJoinOperationId
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinRes

class PlayerActor(private val actorId: String, private val context: ZLinkActorContext) :
    ZLinkActor {
    private val logger = LoggerFactory.getLogger(PlayerActor::class.java)

    var displayName: String = actorId
        private set

    var roomId: String = ""
        private set

    private var pendingRoomId: String? = null
    private val completedJoinOperations = mutableSetOf<ZLinkActorJoinOperationId>()
    var destroyAfterEntrySpotJoin: Boolean = false
        private set

    var disconnected: Boolean = false
        private set

    fun actorId(): String = actorId

    override fun context(): ZLinkActorContext = context

    fun setDisplayName(value: String) {
        displayName = value
    }

    fun joinRoom(value: String) {
        roomId = value
    }

    fun trackDeferredJoin(value: String) {
        check(pendingRoomId == null) { "a room join is already pending" }
        pendingRoomId = value
    }

    override fun onJoinCompleted(completion: ZLinkActorJoinCompletion): CompletionStage<Void> {
        val operationId =
            when (completion) {
                is ZLinkActorJoinCompletion.Accepted -> completion.operationId()
                is ZLinkActorJoinCompletion.Rejected -> completion.operationId()
                is ZLinkActorJoinCompletion.Failed -> completion.operationId()
            }
        if (!completedJoinOperations.add(operationId)) {
            return CompletableFuture.completedFuture(null)
        }

        var matchedRoomId = pendingRoomId
        pendingRoomId = null
        if (completion !is ZLinkActorJoinCompletion.Accepted) {
            return CompletableFuture.completedFuture(null)
        }

        val joined = completion.reply().decode(BingoRoomJoinRes::class.java)
        if (matchedRoomId.isNullOrBlank()) {
            matchedRoomId = joined.state.roomId
        }
        joinRoom(matchedRoomId.orEmpty())
        logger.info("bingo-lifecycle entry-leave actor={}", actorId)
        return CompletableFuture.completedFuture(null)
    }

    fun markForDestroyAfterRoomLeave() {
        destroyAfterEntrySpotJoin = true
    }

    fun markDisconnected() {
        disconnected = true
    }

    fun push(message: Any): CompletionStage<Void> {
        return context.boundSession().send(message).submit()
    }
}
